File: product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        hasZero = False
        howManyZeroes = 0
        product = 1
        
        for i in nums:
            if i == 0:
                hasZero = True
                howManyZeroes += 1
            else:
                product *= i
        if howManyZeroes > 1:
            product = 0 
        for i in range(len(nums)):
            if hasZero == False:
                
                nums[i] = (product//nums[i])
            elif nums[i] != 0 and hasZero and howManyZeroes == 1:
                nums[i] = 0
            elif howManyZeroes > 1:
                nums[i] = 0
            elif nums[i] == 0:
                
                nums[i] = product
        return nums

File: product-of-array-except-self/product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> res(nums.size(), 1);
        
        int prefix = 1;
        for(int i=0; i < nums.size(); i++){
            res.at(i) = prefix;
            prefix *= nums.at(i);
        }
        
        int postfix = 1;
        for(int i=nums.size()-1; i>=0; i--){
            res.at(i) *= postfix;
            postfix *= nums.at(i);
        }
        return res;
    }
};